design/Events_design.md

```diff
@@ -0,0 +1,113 @@
+# Events in Obscuro - Design document
+
+## Scope
+
+This is a design proposal for how to handle events in Obscuro.
+
+## What are events in Ethereum
+
+### Smart contracts can emit events which are streamed from a node and generally used by UIs.
+
+For example:
+
+```emit Transfer(from, to, amount);```
+
+
+### A web app can subscribe to events by doing something like:
+
+```
+var subscription = web3.eth.subscribe('logs', {
+    address: '0x123456..', 
+    topics: ['0x12345...']
+}, function(error, result){
+    if (!error)
+        console.log(result);
+});
+- address - String|Array: An address or a list of addresses to only get logs from particular account(s).
+- topics - Array: An array of values which must each appear in the log entries. The order is important, if you want to leave topics out use null, e.g. [null, '0x00...']. You can also pass another array for each topic with options for that topic e.g. [null, ['option1', 'option2']]
+```
+
+A web app might request:
+- stream all "transfer" events where the 'to' field is my address
+- stream all "transfer" events from the USDC contract
+
+
+### How events work 
+
+The query made on the UI side is transformed in a server-side query on the node, and there is some logic after a tx is executed and events are emmitted, to match them against the filters requested by users and distribute them to the requester.(? - need to confirm this is 100% how it works) .
+
+*Note that there is no constraint on data access, since all data is public.*
+
+
+## Requirements and Problems in Obscuro
+
+- The high level requirement is to determine which account owner can view which events, and make sure, through encryption that nobody else can see those events.
+- Ideally we want to achieve this without any change to the API.
+
+
+#### Problem1: How do we authenticate event subscriptions? 
+We have a wallet extension with viewing key support, which we have to use ideally, to avoid creating another mechanism.
+
+#### Problem2: Assuming the enclave knows who is subscribed, how will the filtering work?
+
+E.g: Alice is requesting to listen to all transfer events. (this is a perfectly reasonable request on a public chain)
+
+There are 2 cases:
```

Comment:
I think we're ignoring an important case here: "The event was emitted by a smart contract as a result of executing a tx sent by Alice, but should not be seen by Alice.". In fact, there are really four cases:

- The event was emitted by a smart contract as a result of executing a tx sent by Alice, and should be seen by Alice
- The event was emitted by a smart contract as a result of executing a tx sent by Alice, but should not be seen by Alice
- The event was emitted by a smart contract as a result of executing a tx sent by Bob, and it should be seen by Alice
- The event was emitted by a smart contract as a result of executing a tx sent by Bob, and it should not be seen by Alice

I think in the rest of the doc, we assume that Alice always has a right to see events caused by transactions she submitted. But is that always true? I'm trying to think of an example - perhaps a winning bid in a gambling contract might emit a message destined for the contract owner _only_ telling them they need to replenish their prize pool?

Even if we decide that Alice gets to see any events caused by her transactions (e.g. because it's simpler), I think we should still acknowledge this as a limitation in the doc.

Reply:
umm. that's a very good point Joel.
wdyt, @CaisManai ?

Reply:
I think this is a moot point @joeldudley, because the events are returned in tx receipts as well, so the person who submitted a tx will always see those events.
I think we can safely make that assumption. I'll state it clearly in the design